components/TemplateGrid/TemplateGridContainer.vue

```diff
@@ -45,202 +45,142 @@ export default {
-        },
-        ghostElement() {
-            // Last row might be 0 - that's why === null
-            return this.lastRow === null ? null : this.gridData[this.lastRow];
+            return this.gridData.filter(element => element.id !== this.ghostElement.id);
         },
         maxRow() {
             return Math.max(...this.dataWithoutGhostElement.map(item => item.row));
         },
     },
+    created() {
+        this.debounceFunc = debounce(this.calculateRowsCount, 200);
+    },
     mounted() {
-        this.calculateRowsCount(0);
-        window.addEventListener('resize', this.calculateRowsCount);
+        this.calculateRowsCount();
+        window.addEventListener('resize', this.debounceFunc);
     },
     destroyed() {
-        window.removeEventListener('resize', this.calculateRowsCount);
+        window.removeEventListener('resize', this.debounceFunc);
     },
     methods: {
-        calculateRowsCount(time = 300) {
-            debounce(() => {
-                const { clientHeight } = document.querySelector('.grid-container');
-                const visibleRows = Math.ceil(clientHeight / this.rowsHeight);
-                const totalRows = Math.max(this.gridData.length, visibleRows) + 1;
-                this.$emit('setRowsCount', { key: 'rowsCount', value: totalRows });
-            }, time)();
+        calculateRowsCount() {
+            const { clientHeight } = document.querySelector('.grid-container');
+            const visibleRows = Math.ceil(clientHeight / this.rowsHeight);
+            const totalRows = Math.max(this.gridData.length, visibleRows) + 1;
+            this.$emit('setRowsCount', totalRows);
         },
-        setGhostItemPosition({ column, row }) {
-            if (
-                (this.lastColumn !== column || this.lastRow !== row)
-                && row !== null
-                && column !== null
-            ) {
-                const newRow = this.setRowIfCollapse(row + 0.5);
-                this.lastColumn = column;
-                this.lastRow = newRow || row;
-                // Seems like lastRow is a ghost element index in the array data
-                // - no need to filter for the ghost element when we have it index
-                this.$emit('addItem', {
-                    item: {
-                        id: this.ghostIndex,
-                        column: this.lastColumn,
-                        row: this.lastRow,
-                        ghost: true,
-                    },
+        onDragOver(event) {
+            const { clientX, clientY } = event;
+            const { overRow, collisionRelation } = this.mousePosition;
+            if (this.onDragFirstItem(event)) return false;
+            const tmpCollisionRelation = this.getMouseOverProps(clientX, clientY);
+            const collidingItem = this.collisionsDetect(overRow);
+
+            if (collidingItem !== null && collisionRelation !== tmpCollisionRelation) {
```

Comment:
conditions are not clear, would use variables to describe them and also I would wrap it into one method